libr/core/visual.c

```diff
@@ -315,42 +315,72 @@ static void prompt_read(const char *p, char *buf, int buflen) {
 	showcursor (NULL, false);
 }
 
-R_API void r_core_visual_prompt_input(RCore *core) {
-	int ret;
-	ut64 addr = core->offset;
-	ut64 bsze = core->blocksize;
-	int h;
-	(void) r_cons_get_size (&h);
-	r_cons_enable_mouse (false);
-	r_cons_gotoxy (0, h - 2);
-	r_cons_reset_colors ();
-	r_cons_printf ("\nPress <enter> to return to Visual mode.\n");
-	r_cons_show_cursor (true);
-	core->vmode = false;
-	ut64 newaddr = addr;
+static void reset_print_cur(RPrint *p) {
+	p->cur = 0;
+	p->ocur = -1;
+}
+
+static void backup_current_addr(RCore *core, ut64 *addr, ut64 *bsze, ut64 *newaddr) {
+	*addr = core->offset;
+	*bsze = core->blocksize;
 	if (core->print->cur_enabled) {
 		if (core->print->ocur != -1) {
 			int newsz = core->print->cur - core->print->ocur;
-			newaddr = core->offset + core->print->ocur;
+			*newaddr = core->offset + core->print->ocur;
 			r_core_block_size (core, newsz);
 		} else {
-			newaddr = core->offset + core->print->cur;
+			*newaddr = core->offset + core->print->cur;
 		}
-		r_core_seek (core, newaddr, 1);
+		r_core_seek (core, *newaddr, 1);
 	}
-	do {
-		ret = r_core_visual_prompt (core);
-		if (core->offset != newaddr) {
-			// do not restore seek anymore
-			newaddr = addr;
+}
+
+static void restore_current_addr(RCore *core, ut64 addr, ut64 bsze, ut64 newaddr) {
+	bool restore_seek = true;
+	if (core->offset != newaddr) {
+		bool cursor_moved = false;
+		// when new address is in the screen bounds, just move
+		// the cursor if enabled and restore seek
+		if (core->print->cur != -1 && core->print->screen_bounds > 1) {
+			if (core->offset >= addr &&
+			    core->offset < core->print->screen_bounds) {
+				core->print->ocur = -1;
+				core->print->cur = core->offset - addr;
+				cursor_moved = true;
+			}
 		}
-	} while (ret);
+
+		if (!cursor_moved) {
+			restore_seek = false;
+			reset_print_cur (core->print);
+		}
+	}
+
 	if (core->print->cur_enabled) {
-		if (addr != newaddr) {
+		if (restore_seek) {
 			r_core_seek (core, addr, 1);
 			r_core_block_size (core, bsze);
 		}
 	}
+}
+
+R_API void r_core_visual_prompt_input(RCore *core) {
+	ut64 addr, bsze, newaddr;
+	int ret, h;
+	(void) r_cons_get_size (&h);
+	r_cons_enable_mouse (false);
+	r_cons_gotoxy (0, h - 2);
+	r_cons_reset_colors ();
+	r_cons_printf ("\nPress <enter> to return to Visual mode.\n");
+	r_cons_show_cursor (true);
+	core->vmode = false;
+
+	backup_current_addr (core, &addr, &bsze, &newaddr);
```

Comment:
I prefer other style for static functions, but anyway, ill merge and we can discuss later on this to specify the coding style rules

Reply:
Which style do you prefer?